crates/bevy_render/src/texture/icon.rs

```diff
@@ -0,0 +1,76 @@
+use crate::texture::Image;
+use bevy_asset::{Assets, Handle};
+use bevy_ecs::{
+    prelude::{Component, Entity, NonSendMut, Query, Res},
+    system::Commands,
+};
+use bevy_log::{error, info};
+use bevy_winit::WinitWindows;
+use winit::window::Icon;
+
+/// An icon that can be placed at the top left of the window.
+#[derive(Component, Debug)]
+pub struct WindowIcon(pub Option<Handle<Image>>);
+
+/// Set or unset the window icon, depending on whether `Some(image_handle)` or `None` is provided.
+///
+/// # Example
+/// ```rust,no_run
+/// use bevy_app::{App, Startup, Update};
+/// use bevy_asset::AssetServer;
+/// use bevy_ecs::prelude::*;
+/// use bevy_render::texture::{set_window_icon, WindowIcon};
+///
+/// fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
+///     let icon_handle = asset_server.load("branding/icon.png");
+///     commands.spawn(WindowIcon(Some(icon_handle)));
+/// }
+///
+/// fn main() {
+///   App::new()
+///     .add_systems(Startup, setup)
+///     .add_systems(Update, set_window_icon)
+///     .run();
+/// }
+/// ```
+///
+/// This functionality [is only known to work on Windows and X11](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.set_window_icon).
+pub fn set_window_icon(
+    images: Res<Assets<Image>>,
+    mut commands: Commands,
+    mut query: Query<(Entity, &mut WindowIcon)>,
+    mut winit_windows: NonSendMut<WinitWindows>,
+) {
+    for (entity, window_icon) in query.iter_mut() {
+        let icon = {
+            if let Some(image) = &window_icon.0 {
+                let Some(icon) = images.get(image) else { continue };
+                let result: Result<Icon, _> = icon.clone().try_into();
+
+                match result {
+                    Ok(icon) => Some(icon),
+                    Err(err) => {
+                        error!("failed to set window icon: {}", err);
+                        commands.entity(entity).remove::<WindowIcon>();
+                        continue;
+                    }
+                }
+            } else {
+                None
+            }
+        };
+
+        if let Some(icon) = &icon {
+            for (_id, window) in &mut winit_windows.windows {
+                window.set_window_icon(Some(icon.clone()));
+            }
+        } else {
+            for (_id, window) in &mut winit_windows.windows {
+                window.set_window_icon(None);
+            }
+        }
+
+        info!("window icon set");
+        commands.entity(entity).remove::<WindowIcon>();
```

Comment:
Definitely don't think we should be removing the Window icon component here once it's set.

Reply:
I was worried about a busy loop of always attempting to check whether an icon exists and then exiting early if it did.  I looked into using the typestate pattern, e.g., `WindowIcon<Loading>`, `WindowIcon<Loaded>`, etc., but ran into difficulties with that.

Is there a way to remove this code from the hot path once the icon has been set?

Reply:
Set a run condition on the system itself, checking if any of the `WindowIcon` components have been changed. That will stop it from running at all, and will be extremely quick to check.

Reply:
As I acquaint myself with Bevy, I'm realizing that the `Changed<WindowIcon>` state will only last a tick, until the next `&mut WindowIcon` occurs, while the image asset itself will take several ticks to load, by which point the `Changed<WindowIcon>` state will have been cleared two or more ticks in the past.  This delay between the `Changed<T>` and the loaded image asset seems to rule out a simple read-only check of some kind, instead requiring a component-specific state machine to track when things are in place to set the icon.

If the state machine is stored on the component itself, mutating the component to update the state enum will interfere with the `Changed<WindowIcon>` check.  And polling in the hot path for whether to set the icon seems to be unavoidable unless loading is carried out in a [separate app state](https://github.com/bevyengine/bevy/blob/caa662272ce91028cef6e237bb7c2a3e78727ef4/examples/2d/texture_atlas.rs#L11-L13).  Using app state seems to have the drawback of not being generic and instead requiring an app-specific setup.

I'm sure this is all just me misunderstanding how one is supposed to do things in Bevy when an asset needs to be fully loaded.  I'll poke around and try to better understand the desired approach in this case.

Reply:
I'm thinking something similar to the implementation for [run_once](https://github.com/bevyengine/bevy/blob/c809779b6e5f00784f66a8694b0b2c3a275b4182/crates/bevy_ecs/src/schedule/condition.rs#L179-L189) might work here.